File: Python/LeetCode/Num17.py
"""
Date:
  23.04.05
Number:
  LeetCode 17
Title:
  Letter Combinations of a Phone Number
Level:
  Medium
Author:
  thelight0804
"""

class Solution(object):
  def letterCombinations(self, digits):
    """
    :type digits: str
    :rtype: List[str]
    """
    result = []

    # set char to phone number
    hashMap = {
      '2': 'abc',
      '3': 'def',
      '4': 'ghi',
      '5': 'jkl',
      '6': 'mno',
      '7': 'pqrs',
      '8': 'tuv',
      '9': 'wxyz'
    }

    """
    index: str
      digits index
    letter: str
      phone char
    """
    def bt(index, letter):
      # exit function
      if len(letter) == len(digits):
        result.append(letter)
        return

      for i in range(index, len(digits)):
        num = digits[i]  # phone number
        chars = hashMap[num]  # chars list
        for j in chars:  # loop char
          bt(i+1, letter+j)  # Recursive Function

    # empty input
    if len(digits) == 0:
      return result

    # start Backtracking
    bt(0, '')

    # output result
    return result

File: JavaScript/LeetCode/Num17.js
/*
Date:
  23.04.26
Number:
  LeetCode 1710
Title:
  Maximum Units on a Truck
Level:
  Easy
Author:
  thelight0804
*/

/**
 * @param {number[][]} boxTypes
 * @param {number} truckSize
 * @return {number}
 */
var maximumUnits = function(boxTypes, truckSize) {
  var units = 0;
  boxTypes.sort((a, b) => b[1] - a[1] ); //sort boxTypes

  //loop box
  for(var i = 0; i < boxTypes.length; i++){
		//truck is full
    if(truckSize <= 0) return units;
		//can put all unit in truck
    if(truckSize > boxTypes[i][0]){
      units += boxTypes[i][0] * boxTypes[i][1]
      truckSize -= boxTypes[i][0]
    }
    else{ //can put part of in truck
      units += boxTypes[i][1] * truckSize;
      truckSize -= boxTypes[i][0]
    }
  }
  return units;
};